### PR TITLE
P.flis/master material

### PR DESCRIPTION
**Pull request by p-flis**
3k samples, 500x281 resolution
matteGlass

### CODE REVIEW

File: RayTracer/Source/Materials/MasterMaterial.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            int sought = 0;
+            for (int i = 0; i < _probabilities.Count; i++)
+            {
+                if (sought < seek && seek <= sought + _probabilities[i])
+                    return i;
+
+                sought += _probabilities[i];
+            }
+
+            return -1;
```
Comment (tomasz-herman) [inline RayTracer/Source/Materials/MasterMaterial.cs:54]: Throw exception with custom message, it will be easier to track

**Review by tomasz-herman — COMMENTED** [review]
Ok, but it needs some setter for ITexture that would propagate ITexure to all materials.
...and other setters that we might need.
It might also contain all 4 materials we implemented so far in hard coded fashion.

File: RayTracer/Source/Materials/MasterMaterial.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        public MasterMaterial(List<IMaterial> materials, List<int> probabilities, AbstractSampler<float> sampler = null)
+        {
+            if (probabilities.Count != materials.Count)
+                throw new ArgumentException("probabilities size != materials size");
+            if (probabilities.Exists(x => x < 0))
+                throw new ArgumentException("probability < 0");
+
+            _materials = materials;
+            _probabilities = probabilities;
+            _sampler = sampler ?? new ThreadSafeSampler<float>(FloatSampling.Random, 10000);
```
Comment (karolk98) [inline RayTracer/Source/Materials/MasterMaterial.cs:24]: We should probably do sth about this magic 10000 at this point ;d

File: RayTracer/Source/Materials/MasterMaterial.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                sought += _probabilities[i];
+            }
+
+            return -1;
+        }
+
+        public bool Scatter(ref Ray ray, ref HitInfo hit, out Color attenuation, out Ray scattered)
+        {
+            var index = GetMaterialIndex();
+            var material = _materials[index];
```
Comment (karolk98) [inline RayTracer/Source/Materials/MasterMaterial.cs:60]: Its a lil weird. Maybe just do GetMaterial functon

**Review by karolk98 — APPROVED** [review]

File: RayTracer/Source/Materials/MasterMaterial.cs (hunk 3 above)
Comment (p-flis) [reply RayTracer/Source/Materials/MasterMaterial.cs]: changed to GetMaterial one

**Comment by p-flis** [comment]
> 
> 
> Ok, but it needs some setter for ITexture that would propagate ITexure to all materials.
> ...and other setters that we might need.
> It might also contain all 4 materials we implemented so far in hard coded fashion.

I've hard-coded materials - it seems to be most appropriate implementation

File: RayTracer/Source/Materials/MasterMaterial.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            if (Reflective != null)
+            {
+                probabilities.Add(ReflectivePart);
+                materials.Add(Reflective);
+            }
+            if (Refractive != null)
+            {
+                probabilities.Add(RefractivePart);
+                materials.Add(Refractive);
+            }
```
Comment (tomasz-herman) [inline RayTracer/Source/Materials/MasterMaterial.cs:49]: It's slow.
Make sure slave materials always exist, add appropriate getters and setters in slave materials, then make slave materials properties get only.

**Comment by tomasz-herman** [comment]
https://github.com/tomasz-herman/RayTracing/pull/39

**Review by tomasz-herman — APPROVED** [review]

**Comment by tomasz-herman** [comment]
This needs some tinkering with shaders before merge.
I started doing some work in this field. 
https://github.com/tomasz-herman/RayTracing/tree/t.herman/master_material_shading_concept
The work above assumes shader will use separate color/texture for ambient/diffuse/specular.

**Pull request closed without merging**